Add a retry_on_timeout decorator and use it on the lowest-level visit step

cworth-gh opened the pull request:
The first commit in this pull request adds the retry_on_timeout decorator needed to deal with TimeoutExceptions generated by code that calls Selenium's set_page_load_timeout, (along with a phantomjs bug where page loads hang forever for some reason).

I've already sent that first commit to upstream in a pull request.

The second commit just adds a little extra print to the log whenever a retry occurs, (so we can watch Travis logs to characterize when the phantomjs bug is acting up).

cworth-gh committed:
Add a retry_on_timeout decorator and use it on the lowest-level visit step

We have seen cases where the PhantomJS webdriver appears to hang
indefinitely. We're not sure what's causing these, but other people
seem to be hitting them as well, see:

	https://github.com/ariya/phantomjs/issues/11526

We have been able to get selenium to raise a TimeoutException by calling:

	browser.set_page_load_timeout(20)  # in seconds

For a complete solution, we need to respond to the TimeoutException by
retrying the page load. This commit adds that try/except block by
means of a new decorator named @retry_on_timeout, (which, other than
which exception is caught, (and not doing any sleeping), is _very_
similar to the existing @wait_for decorator).

cworth-gh committed:
Add some information prints for retry_on_timeout

So that we can see in the logs whenever any retries occur.

cworth-gh committed:
Add an explicit version to setup.py

We can't use use_scm_version since that makes pip-compile unhappy (it
will try to run "python setup.py egg_info" from a checkout with no
.git directory and the use_scm_version code will barf).

Instead, we put an explicit version in place with a local part
indicating the change we have made.

cworth-gh commented:
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;0JUIVEwt&#x2F;676-consider-refreshing-page-when-there-is-an-aloe-testing-timeout"><img src="https:&#x2F;&#x2F;github.trello.services&#x2F;images&#x2F;trello-icon.png" width="12" height="12"> Consider refreshing page when there is an Aloe testing timeout. (2)</a>

ericamador commented on aloe_webdriver/util.py:
```diff
@@ -508,3 +510,35 @@ def wrapped(*args, **kwargs):
+    for (default 5).
+    """
+
+    def wrapped(*args, **kwargs):
```
You can add the functools.wraps decorator to this function to have the convenience of copying the function name, docstring, etc. to the functions you decorate with `@retry_on_timeout`.

https://docs.python.org/2/library/functools.html#functools.wraps

cworth-gh committed:
Add use of @wraps decorator to preserve wrapped function name/docstring

This is applied to both the wait_for and the retry_on_timeout wrappers.

The documentation for the @wraps decorator is here:

    https://docs.python.org/2/library/functools.html#functools.wraps
